src/libraries/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs

```diff
@@ -47,9 +47,35 @@ internal SslAuthenticationOptions(SslServerAuthenticationOptions sslServerAuthen
             TargetHost = string.Empty;
 
             // Server specific options.
-            CertificateRevocationCheckMode = sslServerAuthenticationOptions.CertificateRevocationCheckMode;
-            ServerCertificate = sslServerAuthenticationOptions.ServerCertificate;
             CipherSuitesPolicy = sslServerAuthenticationOptions.CipherSuitesPolicy;
+            CertificateRevocationCheckMode = sslServerAuthenticationOptions.CertificateRevocationCheckMode;
+
+            if (sslServerAuthenticationOptions.ServerCertificateContext != null)
+            {
+                CertificateContext = sslServerAuthenticationOptions.ServerCertificateContext;
+            }
+            else if (sslServerAuthenticationOptions.ServerCertificate != null)
+            {
+                X509Certificate2? certificateWithKey = sslServerAuthenticationOptions.ServerCertificate as X509Certificate2;
+
+                if (certificateWithKey != null && certificateWithKey.HasPrivateKey)
+                {
+                    // given cert is X509Certificate2 with key. We can use it directly.
+                    CertificateContext = SslStreamCertificateContext.Create(certificateWithKey, null);
+                }
+                else
+                {
+                    // This is legacy fix-up. If the Certificate did not have key, we will search stores and we
+                    // will try to find one with matching hash.
+                    certificateWithKey = SecureChannel.FindCertificateWithPrivateKey(this, true, sslServerAuthenticationOptions.ServerCertificate);
+                    if (certificateWithKey == null)
+                    {
+                        throw new AuthenticationException(SR.net_ssl_io_no_server_cert);
+                    }
+
+                    CertificateContext = new SslStreamCertificateContext(certificateWithKey);
```

Comment:
```suggestion
                    CertificateContext = SslStreamCertificateContext.Create(certificateWithKey, null);
```

Reply:
I added internal `Create()` to be more consistent but allowing to have SslStreamCertificateContext without building full chain to support legacy behavior on platform where we do not build chain but leave OS to do it.